src/test/java/com/palantir/common/streams/BackpressureSpliteratorTests.java

```diff
@@ -0,0 +1,125 @@
+        }).thenAnswer(x -> {
+            Consumer<CompletableFuture<String>> consumer = x.getArgument(0);
+            consumer.accept(otherFuture);
+            return true;
+        }).thenReturn(false);
+    }
+
+    @Test
+    public void returnsFalseWhenAllFuturesCompleted() {
+        Spliterator<String> spliterator = new BackpressureSpliterator<>(Stream.<CompletableFuture<String>>empty().spliterator(), 1);
+        assertThat(spliterator.tryAdvance(consumer)).isFalse();
+        verifyZeroInteractions(consumer);
+    }
+
+    @Test
+    public void throwsIfSuppliedFutureThrows() {
+        CompletableFuture<String> someFuture = new CompletableFuture<>();
+        Spliterator<String> spliterator =
+                new BackpressureSpliterator<>(Stream.of(someFuture).spliterator(), 1);
+        someFuture.completeExceptionally(new RuntimeException());
+        assertThatExceptionOfType(CompletionException.class).isThrownBy(() -> spliterator.tryAdvance(consumer));
+    }
+
+    @Test
+    public void onlyRunsUpToDesiredConcurrencyTasksSimultaneously() {
+        Spliterator<String> spliterator = new BackpressureSpliterator<>(sourceSpliterator, 1);
+
+        String firstData = "firstData";
+        future.complete(firstData);
+        assertThat(spliterator.tryAdvance(consumer)).isTrue();
+        verify(sourceSpliterator, times(1)).tryAdvance(any());
+        verify(consumer).accept(firstData);
+
+        String secondData = "secondData";
+        otherFuture.complete(secondData);
+        assertThat(spliterator.tryAdvance(consumer)).isTrue();
+        verify(consumer).accept(secondData);
+
+        assertThat(spliterator.tryAdvance(consumer)).isFalse();
+        verifyNoMoreInteractions(consumer);
+    }
+
+    @Test
+    public void runsDesiredConcurrencyTasksSimultaneously() {
+        future.complete("some string");
+        new BackpressureSpliterator<>(sourceSpliterator, 2).tryAdvance(consumer);
+
+        verify(sourceSpliterator, times(2)).tryAdvance(any());
+    }
+
+    @Test
+    public void doesNotStartNextTaskUntilDoneWithLastValue() {
+        future.complete("some string");
+        Spliterator<String> spliterator = new BackpressureSpliterator<>(sourceSpliterator, 1);
+
+        future.complete("data");
+        spliterator.tryAdvance(consumer);
+
+        verify(sourceSpliterator, times(1)).tryAdvance(any());
+    }
+
+    // This test exists because of an implementation bug while writing this.
+    @Test
+    public void canHandleFutureAlreadyCompleted() {
+        String data = "data";
+        CompletableFuture<String> someFuture = CompletableFuture.completedFuture(data);
+
+        Spliterator<String> spliterator = new BackpressureSpliterator<>(Stream.of(someFuture).spliterator(), 1);
+
+        assertThat(spliterator.tryAdvance(consumer)).isTrue();
+        verify(consumer).accept(data);
+        assertThat(spliterator.tryAdvance(consumer)).isFalse();
+    }
+
+    @Test
+    public void testEstimateSize_hasSize() {
```

Comment:
Rewrite this to iterate through the elements one by one and check the estimate is one lower each time, starting at 5 before you consume any elements, and finally returning 0

Reply:
done.